### PR TITLE
[DATA-3338] - collector test improvements

### PR DESCRIPTION
**Pull request by nicksanford**
[Ticket](https://viam.atlassian.net/browse/DATA-3338) 

This test refactor doesn't change any behavior of the data manager or collectors; it just tests collectors in a way that will make the upcoming PR that fixes the vision.CaptureAllFromCamera and camera.GetImages collectors smaller.

1. Add tests for slam & camera collectors
2. Change to a uniform way of describing the expected sensor proto values in collector tests. This makes it more obvious what the actually tabular data stored in proto files looks like.
3. Change collector tests to not rely on an arbitrary number of retries, making them more reliable.
4. Use actual binary data in collectors that handle binary data

### CODE REVIEW

File: testutils/file_utils.go [hunk 1]
```diff
@@ -100,33 +102,79 @@ func BuildTempModuleWithFirstRun(tb testing.TB, modDir string) string {
...
+	select {
+	case m.Writes <- item:
+	case <-m.ctx.Done():
+	}
 	return nil
 }
 
 // Flush does nothing in this implementation as all data will be stored in memory.
 func (m *MockBuffer) Flush() error {
-	m.lock.Lock()
```
Comment (nicksanford) [inline testutils/file_utils.go:117]: this did nothing

File: testutils/file_utils.go [hunk 2]
```diff
@@ -100,33 +102,79 @@ func BuildTempModuleWithFirstRun(tb testing.TB, modDir string) string {
...
+	return &MockBuffer{
+		ctx:    ctx,
+		Writes: make(chan *v1.SensorData, 1),
+	}
+}
+
+// CheckMockBufferWrites checks that the writes to either
+// MockBuffer.BinaryWrites or MockBuffer.TabularWrites
+// are match the expected data & metadata (timestamps).
+func CheckMockBufferWrites(
```
Comment (nicksanford) [inline testutils/file_utils.go:119]: consider moving this to a method on MockBuffer

File: testutils/file_utils.go [hunk 3]
```diff
@@ -100,33 +103,99 @@ func BuildTempModuleWithFirstRun(tb testing.TB, modDir string) string {
...
+		t.Error("timeout")
+		t.FailNow()
+	case write := <-writes:
+		end := time.Now()
+		// nil out to make comparable
+		requestedAt := write.Metadata.TimeRequested.AsTime()
+		receivedAt := write.Metadata.TimeReceived.AsTime()
+		test.That(t, start, test.ShouldHappenOnOrBefore, requestedAt)
+		test.That(t, requestedAt, test.ShouldHappenOnOrBefore, receivedAt)
+		test.That(t, requestedAt, test.ShouldHappenOnOrBefore, receivedAt)
```
Comment (n0nick) [inline testutils/file_utils.go:162]: duplicate line

File: testutils/file_utils.go [hunk 4]
```diff
@@ -100,33 +103,99 @@ func BuildTempModuleWithFirstRun(tb testing.TB, modDir string) string {
...
 type MockBuffer struct {
-	lock   sync.Mutex
-	Writes []*v1.SensorData
+	ctx    context.Context
+	cancel context.CancelFunc
+	Writes chan *v1.SensorData
 }
 
-// Write adds a collected sensor reading to the array.
+// NewMockBuffer returns a mock buffer.
```
Comment (n0nick) [inline testutils/file_utils.go:111]: consider documenting this needs to be `Close()` _before_ the collector closes

File: components/camera/collectors_test.go [hunk 5]
```diff
@@ -0,0 +1,208 @@
...
+		return nil, err
+	}
+	return anyVal, nil
+}
+
+func TestCollectors(t *testing.T) {
+	logger := logging.NewTestLogger(t)
+	methodParams, err := convertStringMapToAnyPBMap(map[string]string{"camera_name": "camera-1", "mime_type": "image/jpeg"})
+	test.That(t, err, test.ShouldBeNil)
+	viamLogoJpeg, err := io.ReadAll(base64.NewDecoder(base64.StdEncoding, bytes.NewReader(viamLogoJpegB64)))
```
Comment (n0nick) [inline components/camera/collectors_test.go:79]: q, is it actually important this is a valid jpeg, or can this be an arbitrary byte array?

File: components/camera/collectors_test.go [hunk 6]
```diff
@@ -0,0 +1,208 @@
...
+
+	pcd, err := pointcloud.NewFromFile(artifact.MustPath("pointcloud/test.las"), logger)
+	test.That(t, err, test.ShouldBeNil)
+
+	var pcdBuf bytes.Buffer
+	test.That(t, pointcloud.ToPCD(pcd, &pcdBuf, pointcloud.PCDBinary), test.ShouldBeNil)
+
+	now := time.Now()
+	nowPB := timestamppb.New(now)
+	cam := newCamera(img, img, now, pcd)
```
Comment (n0nick) [inline components/camera/collectors_test.go:99]: you always test on a camera where `colorImage`==`depthImage`, how about we have 2 different images for completeness? (if it's easier, one can be the reversed bytes of the other 🤷)

File: components/camera/collectors_test.go (hunk 5 above)
Comment (n0nick) [reply components/camera/collectors_test.go:79]: same question about `pcd`

**Review by n0nick — APPROVED** [review]
LGTM otherwise 💪

File: components/camera/collectors_test.go (hunk 5 above)
Comment (nicksanford) [reply components/camera/collectors_test.go:79]: It is necessary for the images & pointclouds to be valid yes. This is b/c the Go SDK for camera (which the collector calls to get the data it is trying to capture):
https://github.com/viamrobotics/rdk/pull/4551/files/8694045a25f71f67bd774cebee5eec50889a63e8#diff-dcdd8001ea2dc2e271d7bbbb14f98041af0e66ffb025455461a246140cc0c5a3R182-R208

returns `image.Image` (part of the go standard library) and `pointcloud.Pointcloud` a type we define in rdk.

Technically speaking, we could shove invalid data into an image.Image, but I think it is less work to just test how we expect the software to work.

Comment (n0nick) [reply components/camera/collectors_test.go:79]: gotcha, thank you for explaining 👍

File: components/camera/collectors_test.go (hunk 6 above)
Comment (n0nick) [reply components/camera/collectors_test.go:99]: ack! reversing the bytes was in re my suggestion to use just arbitrary bytes as images, but this issue makes sense as you've explained we assume an `image.Image` in the camera SDK.

File: testutils/file_utils.go (hunk 4 above)
Comment (nicksanford) [reply testutils/file_utils.go:111]: done

File: testutils/file_utils.go (hunk 3 above)
Comment (nicksanford) [reply testutils/file_utils.go]: done

**Pull request merged**